csrc/quantization/quant_llm/quant_llm_linear.cu

```diff
@@ -0,0 +1,302 @@
+//    Copyright 2024 FP6-LLM authors
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+// This file is adapted from
+// https://github.com/usyd-fsalab/fp6_llm/blob/5df6737cca32f604e957e3f63f03ccc2e4d1df0d/fp6_llm/csrc/fp6_linear.cu
+
+#include "quant_matmul.cuh"
+#include "quant_reduction.cuh"
+
+#include <stdio.h>
+#include <assert.h>
+
+namespace vllm {
+
+template <typename TilingConfig, typename OutputDataType, int EXPONENT,
+          int MANTISSA>
+static void Kernel_Ex(cudaStream_t stream, const uint4* Weight,
+                      const half* Scales, const half* B, OutputDataType* C,
+                      const size_t M_Global, const size_t N_Global,
+                      const size_t K_Global, int Split_K) {
+#ifdef DEBUG_MODE
+  printf("\n");
+  printf("Launcher.cu->Kernel_Ex():\n");
+  printf("M: %d, N: %d, K: %d, SplitK: %d\n", M_Global, N_Global, K_Global,
+         Split_K);
+  printf("TILE_M: %d, TILE_K: %d, TILE_N: %d\n", TilingConfig::TILE_M,
+         TilingConfig::TILE_K, TilingConfig::TILE_N);
+#endif
+  static size_t SHMEM_SZ =
+      max(TilingConfig::SMEM_SIZE_B_TILE + SMEM_SIZE_PER_TB_A_TILE,
+          TilingConfig::SMEM_SIZE_C_TILE);
+  cudaFuncSetAttribute(
+      QUANT_GEMM_Kernel<TilingConfig, OutputDataType, EXPONENT, MANTISSA>,
+      cudaFuncAttributeMaxDynamicSharedMemorySize, SHMEM_SZ);
+  size_t dimN = (N_Global - 1) / TilingConfig::TILE_N + 1;
+  size_t dimM = M_Global * Split_K / TilingConfig::TILE_M;
+  dim3 GridDim(dimN, dimM, 1);
+  dim3 BlockDim(WARP_SIZE * TilingConfig::BLOCK_WARPS, 1, 1);
+//
+#ifdef DEBUG_MODE
+  printf(
+      "GridDim.x: %d, GridDim.y: %d, GridDim.z: %d, BlockDim.x: %d, "
+      "BlockDim.y: %d, BlockDim.z: %d SHMEM_SZ: %d\n",
+      GridDim.x, GridDim.y, GridDim.z, BlockDim.x, BlockDim.y, BlockDim.z,
+      SHMEM_SZ);
+  printf("\n");
+#endif
+  QUANT_GEMM_Kernel<TilingConfig, OutputDataType, EXPONENT, MANTISSA>
+      <<<GridDim, BlockDim, SHMEM_SZ, stream>>>(Weight, Scales, B, C, M_Global,
+                                                N_Global, K_Global, Split_K);
+}
+
+template <int EXPONENT, int MANTISSA>
+cudaError_t fpx_linear_kernel(
```

Comment:
I'm a bit worried about all the generated functions here contributing to the binary size. I count 17 instantiations of this function and then the 10 (12?) kernel instantiations in this function, presumably producing ~170 implementations. We should definitely measure the increase cc @tlrmchlsmth

Reply:
I was worried about this as well, but when I built it with way more instantiations (over 50), I didn't notice any significant increase in wheel size. It's overall a pretty small kernel when built.